File: 0198-house-robber/0198-house-robber.go
func rob(nums []int) int {

	if len(nums) < 2 {
		return nums[0]
	}

	dp := make([]int, len(nums))
	dp[0] = nums[0]
	if nums[1] > nums[0] {
		dp[1] = nums[1]
	} else {
		dp[1] = nums[0]
	}

	for i := 2 ; i< len(nums) ; i++ {
		if dp[i - 1] > nums[i] + dp[i  - 2] {
			dp[i] = dp[i - 1]
		} else {
			dp[i] = nums[i] + dp[i-2]
		}
	}
	return dp[len(nums) - 1]
}

File: leetcode/0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: list[int]) -> int:
        N = len(nums)
        if N < 2:
            return nums[0]
        dp = [0] * N
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])
        for n in range(2, N):
            dp[n] = max(nums[n] + dp[n - 2], dp[n - 1])
        return dp[-1]